From 184735c825dc7fb29dfb91294502129f50478bc6 Mon Sep 17 00:00:00 2001
From: Eric Sauvageau <merlin@asuswrt-merlin.net>
Date: Sun, 28 Jun 2026 21:56:34 -0400
Subject: [PATCH] rc: webui: rework patch that allows user-defined IPv6 WAN
 prefix

The original patch was breaking IPv6 functionality in SDN/GNP.

Changes over the original patch:

- Do not hardcode 64 into ipv6_prefix_length nvram, as this breaks
  the ability to subnet a /64 to guest networks/SDN whenever you get
  a /56 from your ISP.

- A prefix has changed only if received size doesn't match the
  previously received size (ipv6_prefix_length), not if it doesn't match
  the user requested size (ipv6_prefix_len_wan).

- Don't overwrite ipv6_prefix_len_wan if a prefix has changed (or
  in the original patch if it didn't match the requested size).
  Leave user inputs untouched to avoid confusion over user settings
  changed on their own.

(cherry picked from commit 61983c0fa4b7c0fa62a991028de0ae1ea006398e)
---
 release/src/router/rc/udhcpc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/release/src/router/rc/udhcpc.c b/release/src/router/rc/udhcpc.c
index 7a8c51c652..6a79e28822 100644
--- a/release/src/router/rc/udhcpc.c
+++ b/release/src/router/rc/udhcpc.c
@@ -1966,15 +1966,14 @@ bound6(char *wan_ifname, int bound)
 			goto skip;
 
 		prefix_changed = (!nvram_match(ipv6_nvname_by_unit("ipv6_prefix", wan_unit), addr) ||
-					nvram_get_int(ipv6_nvname_by_unit("ipv6_prefix_len_wan", wan_unit)) != size);
+					nvram_get_int(ipv6_nvname_by_unit("ipv6_prefix_length", wan_unit)) != size);
 		if (prefix_changed) {
 			eval("ip", "-6", "addr", "flush", "scope", "global", "dev", lan_ifname);
 			nvram_set(ipv6_nvname_by_unit("ipv6_rtr_addr", wan_unit), "");
 			nvram_set(ipv6_nvname_by_unit("ipv6_prefix", wan_unit), addr);
-			nvram_set_int(ipv6_nvname_by_unit("ipv6_prefix_length", wan_unit), 64);
+			nvram_set_int(ipv6_nvname_by_unit("ipv6_prefix_length", wan_unit), size);
 			logmessage("dhcp6 client", "WAN Prefix Size Requested:/%d, Received:/%d",
 				 nvram_get_int(ipv6_nvname_by_unit("ipv6_prefix_len_wan", wan_unit)), size);
-			nvram_set_int(ipv6_nvname_by_unit("ipv6_prefix_len_wan", wan_unit), size);
 		}
 		if (*addr)
 			add_ip6_lanaddr();
-- 
2.25.1

